File: java/src/main/java/com/edu/agent/agent/AssessmentAgent.java
package com.edu.agent.agent;

import com.edu.agent.core.AgentEvent;
import com.edu.agent.core.EventType;
import com.edu.agent.core.LearnerModel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.context.event.EventListener;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Assessment Agent -- BKT 知识追踪评估。
 * <p>
 * 使用 Spring @EventListener + @Async 实现异步事件响应。
 * 对应Python版的 assessment_agent.py。
 */
@Component
public class AssessmentAgent {

    private static final Logger log = LoggerFactory.getLogger(AssessmentAgent.class);
    private final ApplicationEventPublisher publisher;
    private final Map<String, LearnerModel> learnerModels = new ConcurrentHashMap<>();

    public AssessmentAgent(ApplicationEventPublisher publisher) {
        this.publisher = publisher;
    }

    public Map<String, LearnerModel> getLearnerModels() { return learnerModels; }

    @Async
    @EventListener(condition = "#event.type == T(com.edu.agent.core.EventType).STUDENT_SUBMISSION")
    public void handleSubmission(AgentEvent event) {
        String learnerId = event.getLearnerId();
        String knowledgeId = (String) event.getData().getOrDefault("knowledge_id", "");
        boolean isCorrect = (boolean) event.getData().getOrDefault("is_correct", false);

        LearnerModel model = learnerModels.computeIfAbsent(learnerId, LearnerModel::new);
        LearnerModel.KnowledgeState state = model.updateMastery(knowledgeId, isCorrect);

        log.info("[Assessment] learner={}, kp={}, correct={}, mastery={:.3f} ({})",
                learnerId, knowledgeId, isCorrect, state.getMastery(), state.getLevel());

        // Publish mastery updated
        Map<String, Object> data = new HashMap<>();
        data.put("knowledge_id", knowledgeId);
        data.put("mastery", state.getMastery());
        data.put("level", state.getLevel());
        data.put("is_correct", isCorrect);
        data.put("attempts", state.getAttempts());
        publisher.publishEvent(new AgentEvent(this, EventType.MASTERY_UPDATED,
                "AssessmentAgent", learnerId, data));

        if (state.getMastery() < 0.3 && state.getAttempts() >= 3) {
            Map<String, Object> weakData = new HashMap<>();
            weakData.put("knowledge_id", knowledgeId);
            weakData.put("mastery", state.getMastery());
            publisher.publishEvent(new AgentEvent(this, EventType.WEAKNESS_DETECTED,
                    "AssessmentAgent", learnerId, weakData));
        }

        // Publish assessment complete
        Map<String, Object> completeData = new HashMap<>(data);
        completeData.put("overall_progress", model.getOverallProgress());
        publisher.publishEvent(new AgentEvent(this, EventType.ASSESSMENT_COMPLETE,
                "AssessmentAgent", learnerId, completeData));
    }
}

File: python/agents/assessment_agent.py
"""
Assessment Agent（评估Agent）-- 知识点掌握度评估与学习路径诊断。

核心职责：
1. 接收学生答题结果，用BKT算法更新mastery
2. 检测薄弱知识点，通知Curriculum Agent调整路径
3. 定期生成学习报告

面试要点：
- 为什么用BKT而不是简单的正确率？
  正确率是静态的，BKT可以追踪动态学习过程，考虑猜测和失误
- 为什么用Beta分布表示mastery？
  Beta分布是二项分布的共轭先验，更新方便，能表达不确定性
"""

import logging

from .base_agent import BaseAgent
from core.event_bus import Event, EventType

logger = logging.getLogger(__name__)


class AssessmentAgent(BaseAgent):
    """评估Agent：实时追踪学生对每个知识点的掌握程度。"""

    @property
    def subscribed_events(self) -> list[EventType]:
        return [
            EventType.STUDENT_SUBMISSION,
            EventType.STUDENT_QUESTION,
        ]

    async def handle_event(self, event: Event) -> None:
        if event.type == EventType.STUDENT_SUBMISSION:
            await self._handle_submission(event)
        elif event.type == EventType.STUDENT_QUESTION:
            await self._handle_question(event)

    async def _handle_submission(self, event: Event) -> None:
        """
        处理学生答题提交。

        event.data 结构：
        {
            "knowledge_id": "quadratic_eq",
            "is_correct": true,
            "answer": "x = 2",
            "time_spent_seconds": 45
        }
        """
        learner_id = event.learner_id
        data = event.data
        knowledge_id = data.get("knowledge_id", "")
        is_correct = data.get("is_correct", False)

        model = self.get_learner_model(learner_id)
        state = model.update_mastery(knowledge_id, is_correct)

        await self.emit(
            EventType.MASTERY_UPDATED,
            learner_id,
            {
                "knowledge_id": knowledge_id,
                "mastery": state.mastery,
                "level": state.level.value,
                "attempts": state.attempts,
                "streak": state.streak,
                "confidence": state.confidence,
            },
        )

        if state.mastery < 0.3 and state.attempts >= 3:
            await self.emit(
                EventType.WEAKNESS_DETECTED,
                learner_id,
                {
                    "knowledge_id": knowledge_id,
                    "mastery": state.mastery,
                    "attempts": state.attempts,
                    "suggestion": "需要回到前置知识点复习",
                },
            )

        await self.emit(
            EventType.ASSESSMENT_COMPLETE,
            learner_id,
            {
                "knowledge_id": knowledge_id,
                "is_correct": is_correct,
                "mastery": state.mastery,
                "level": state.level.value,
                "overall_progress": model.get_overall_progress(),
            },
        )

    async def _handle_question(self, event: Event) -> None:
        """处理学生提问，分析涉及的知识点。"""
        learner_id = event.learner_id
        question = event.data.get("question", "")
        knowledge_id = event.data.get("knowledge_id", "general")

        model = self.get_learner_model(learner_id)
        state = model.get_state(knowledge_id)

        await self.emit(
            EventType.ASSESSMENT_COMPLETE,
            learner_id,
            {
                "knowledge_id": knowledge_id,
                "question": question,
                "current_mastery": state.mastery,
                "current_level": state.level.value,
                "context": "student_question",
            },
        )
